src/Tempest/Http/src/RouteConfig.php

```diff
@@ -40,25 +52,24 @@ public function addRoute(MethodReflector $handler, Route $route): self
         return $this;
     }
 
-    /**
-     * Build one big regex for matching request URIs.
-     * See https://github.com/tempestphp/tempest-framework/pull/175 for the details
-     */
-    private function addToMatchingRegex(Route $route, string $routeMark): void
+    public function prepareMatchingRegexes(): void
     {
-        // Each route, say "/posts/{postId}", which would have the regex "/posts/[^/]+", is marked.
-        // e.g "/posts/[^/]+ (*MARK:a)".
-        // This mark can then be used to find the matched route via a hashmap-lookup.
-        $routeRegexPart = "{$route->matchingRegex} (*" . GenericRouter::REGEX_MARK_TOKEN . ":{$routeMark})";
-
-        if (! array_key_exists($route->method->value, $this->matchingRegexes)) {
-            // initialize matching regex for method
-            $this->matchingRegexes[$route->method->value] = "#^(?|{$routeRegexPart})$#x";
-
+        if (! empty($this->matchingRegexes)) {
             return;
         }
 
-        // insert regex part of this route into the matching group of the regex for the method
-        $this->matchingRegexes[$route->method->value] = substr_replace($this->matchingRegexes[$route->method->value], "|{$routeRegexPart}", -4, 0);
+        $this->matchingRegexes = $this->routingTree->toMatchingRegexes();
+    }
+
+    public function __sleep(): array
+    {
+        $this->prepareMatchingRegexes();
+
+        return ['staticRoutes', 'dynamicRoutes', 'matchingRegexes'];
+    }
+
+    public function __wakeup(): void
+    {
+        $this->routingTree = new RoutingTree();
     }
```

Comment:
These magic methods are actually used during `serialize` and `unserialize` I opted to use these instead of `__serialize` and `__unserialize` since those require you to return arrays and serialize everything yourself. This `__sleep` call instruct which properties to serialize (and to ensure `matchingRegexes` is filled by calling `prepareMatchingRegexes). The `__wakeup` call is called after `unserialize`  ensure your internal state is consistent.  In my case I just initialize an empty routing tree, without it this routing tree would be undefined since `unserialize` doesn't go through our normal constructor and since I do not serialize it in the `__sleep` call, I do not need it since we got the end product `matchingRegexes` already.

Reply:
Fair enough, maybe this explanation should be added as a comment, because I'll admit I was a bit confused why we created an empty routing tree at first.

Reply:
done